peer/brontide.go

```diff
@@ -3653,18 +3648,50 @@ func (p *Brontide) handleCloseMsg(msg *closeMsg) {
 			chanCloser.CloseRequest().Err <- err
 		}
 		delete(p.activeChanCloses, msg.cid)
-		return
 	}
 
-	// Queue any messages to the remote peer that need to be sent as a part of
-	// this latest round of negotiations.
-	for _, msg := range msgs {
-		p.queueMsg(msg, nil)
+	// Next, we'll process the next message using the target state machine.
+	// We'll either continue negotiation, or halt.
+	switch typed := msg.msg.(type) {
```

Comment:
Drive by review (was just sent this commit in isolation), but I think that we should move this back into something _like_ `ProcessCloseMsg`. Then this way the peer doesn't need to be aware of the messages it needs to route to the chan closer _and_ which methods to call in order to execute them. 

Then in then in the future when we add the new simple co-op close state machine, this call site stays mostly the same (we break out a new interface to abstract away from the protocol being executed). 

If we want to keep the state machine mostly pure, we can have it emit an event (similar to above), and there's a certain event type that says: "send this message to the peer". Otherwise, the chan closer can gain a slim `Peer` interface, that's just used to flush out new messages. Unit tests can have this send to a channel, while the actual system passes in the peer directly (implicit interface).

Reply:
After trying to rebundle the ChanCloser APIs I decided it was better to leave them separate for the time being.

Reply:
> I also generally oppose the philosophy of only implementing infrastructure when there's overwhelming demand for it

I'm a bit hesitant to this approach because we can end up with code that's generalized, but is only used by one caller.

> Wouldn't this defy the goals of "not putting shutdown logic in the link"?

I think this would be a lot slimmer than what I wrote in #6760, especially since the link isn't receiving/sending out `Shutdown` in the link code and instead uses a callback. It would have some shutdown logic (atomic booleans), but I think it would be minimal. It would also let us do away with the `switch` case in the peer code. Thoughts @Roasbeef ?

Reply:
> I'm a bit hesitant to this approach because we can end up with code that's generalized, but is only used by one caller.

That is the failure mode, yes. I guess I just don't see that as an issue in practice. In fact, generalization has benefits beyond de-duplicating code. The more general the API, the more restrictive the implementation and the more restrictive the implementation, the fewer opportunities there are for implementing it incorrectly. It simplifies the properties of the system. I'll admit that the drawback is that it often takes longer to come up with good generalizations and that effort may not be worth what you get in the short term.

Reply:
> I think this would be a lot slimmer than what I wrote in https://github.com/lightningnetwork/lnd/pull/6760, especially since the link isn't receiving/sending out Shutdown in the link code and instead uses a callback. It would have some shutdown logic (atomic booleans), but I think it would be minimal. It would also let us do away with the switch case in the peer code. Thoughts @Roasbeef ?

After reviewing [option_simple_close](https://github.com/lightning/bolts/pull/1096/files) it is clear to me that `shutdown` is fundamentally a link operation. Conceptually this seems obvious to me now, as any coop-close flow relies on clean channel state and regardless of what protocol you use for coming to agreement on a coop-close transaction, that state machine doesn't really kick off until the link shuts down. We can confirm this hypothesis by reviewing `option_simple_close` and see that it begins the same way:

1. exchange shutdown messages
2. wait for htlcs to drain
3. begin negotiation
4. ?????
5. profit

So if we want to ditch the lifecycle hook work in this PR I am OK with that so long as we acknowledge that `shutdown` truly is part of the Link's lifecycle. The more I think about it, the more I think this is the right thing to do irrespective of our opinion on lifecycle hooks and premature generalization. You could imagine a world with multiple choices for closing negotiation protocols and that you might select which one you want to use *in the shutdown message*. It wouldn't make sense to do so before the `shutdown` unless it was pre-negotiated up front.

`shutdown` is the happy-path end of the link's lifecycle, NOT the beginning of the closing protocol.

Thank you for comming to my TED Talk. 🤪

Reply:
4 days later my conviction in the above has only grown. I think that we should make `shutdown` a message that the link processes, and pull it out of the `ChanCloser`

Reply:
>  I think that we should make shutdown a message that the link processes, and pull it out of the ChanCloser

FWIW with the new co-op close flow, `shutdown` can be sent again to reset state to RBF a new co-op close offer. In this state, the link no longer exists as it isn't loaded into the switch on connection establishment (channel state is now borked).

Reply:
I generally agree tho, that we can just literally move this into what used to be `ProcessCloseMsg` just hiding some of the internal message routing details from the peer itself, and this are mostly the same. 

Then this way it also doesn't need to understand the new dispatch routing for the new co-op close state machine (there're two new messages added).

Reply:
Ok theres one final problem with rebundling that I believe is fatal. We can't escape inspecting the close message type because we need to decide whether or not to register the flush hook only if the message type is a shutdown. If we receive a valid closing_signed then we are already flushed, and shouldn't be flushing again.